sstables/generation_type.hh

```diff
@@ -9,9 +9,70 @@
 #pragma once
 
 #include <cstdint>
+#include <concepts>
+#include <compare>
+#include <limits>
+#include <iostream>
+#include <seastar/core/sstring.hh>
 
 namespace sstables {
-using generation_type = int64_t;
-constexpr int64_t generation_value(generation_type generation) { return generation; }
-constexpr generation_type generation_from_value(int64_t value) { return value; }
+
+template <typename V>
+requires std::copyable<V> && std::equality_comparable<V> && std::three_way_comparable<V>
+struct generation_wrapper {
+    using value_type = V;
+private:
+    V _value;
+public:
+    generation_wrapper() = delete;
+
+    explicit constexpr generation_wrapper(const V& value) noexcept: _value(value) {}
+    constexpr const V& value() const noexcept { return _value; }
+
+    constexpr bool operator==(const generation_wrapper& other) const noexcept { return _value == other._value; }
+    constexpr std::strong_ordering operator<=>(const generation_wrapper& other) const noexcept { return _value <=> other._value; }
+};
+
+using generation_type = generation_wrapper<int64_t>;
```

Comment:
Why a template?

Reply:
a couple of speculative reasons:
- with a template, changing over wholesale from `int64_t` to (say) `time_uuid` requires just changing the `generation_type` typedef and the `generation_from_value` function
- if instead it is decided that `generation_type` should support both representations (converting the sstable unit tests looks like a major project, for example), you'll be able to define it as a wrapper around `std::variant<generation_wrapper<int64_t>,generation_wrapper<time_uuid>>`

I don't know what @psarna plans exactly, maybe he'd like to weigh in.

(in any case I don't feel very strongly about the templated implementation, it may well be overkill)

Reply:
Why do we even need to expose the underlying type in the wrapper? Can it be just:
```C++
struct generation_type {
    some_actual_type value;
    // convenience members go here
};
```

Reply:
This can be changed later just the same.

Reply:
We can't change the type due to compatibility reasons, only add more alternatives (e.g. it will be a variant).

Better to drop the template, it's just obfuscation.

Reply:
Also, let's just call it generation_type rather than generation_wrapper, _wrapper gives the feeling of something useless and artificial, and while it's maybe correct here, let's not depress readers.

Reply:
`generation_wrapper` is the template name, the concrete type (`generation_wrapper<int64_t>`) is still `generation_type`.   but I see the general sentiment in this thread :)

Reply:
btw I don't much like `generation_type` either, can't it be just `generation`?  it's clear from context that it's a type.